google-cloud-firestore/src/main/java/com/google/cloud/firestore/FirestoreBundle.java

```diff
@@ -0,0 +1,228 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.firestore;
+
+import com.google.api.client.util.Lists;
+import com.google.cloud.Timestamp;
+import com.google.firestore.proto.BundleElement;
+import com.google.firestore.proto.BundleMetadata;
+import com.google.firestore.proto.BundledDocumentMetadata;
+import com.google.firestore.proto.BundledQuery;
+import com.google.firestore.proto.NamedQuery;
+import com.google.firestore.v1.Document;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.util.JsonFormat;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/** Represents a Firestore data bundle with results from the given document and query snapshots. */
+public final class FirestoreBundle {
+
+  static int BUNDLE_SCHEMA_VERSION = 1;
+
+  // Raw byte array to hold the content of the bundle.
+  private byte[] bundleData;
+
+  /** Builds a Firestore data bundle with results from the given document and query snapshots. */
+  public static final class Builder {
+    // Id of the bundle.
+    private String id;
+    // Resulting documents for the bundle, keyed by full document path.
+    private Map<String, BundledDocument> documents = new HashMap<>();
+    // Named queries saved in the bundle, keyed by query name.
+    private Map<String, NamedQuery> namedQueries = new HashMap<>();
+    // The latest read time among all bundled documents and queries.
+    private Timestamp latestReadTime = Timestamp.MIN_VALUE;
+    // Printer to encode protobuf objects into JSON string.
+    private JsonFormat.Printer printer = JsonFormat.printer();
+
+    public Builder(String id) {
+      this.id = id;
+    }
+
+    /**
+     * Adds a Firestore document snapshot to the bundle. Both the documents data and the document
+     * read time will be included in the bundle.
+     *
+     * @param documentSnapshot A document snapshot to add.
+     * @returns This instance.
+     */
+    public Builder add(DocumentSnapshot documentSnapshot) {
+      return add(documentSnapshot, Optional.empty());
+    }
+
+    private Builder add(DocumentSnapshot documentSnapshot, Optional<String> queryName) {
+      String documentName = documentSnapshot.getReference().getName();
+      BundledDocument originalDocument = documents.get(documentSnapshot.getReference().getName());
+      List<String> queries =
+          originalDocument == null
+              ? Lists.newArrayList()
+              : Lists.newArrayList(originalDocument.getMetadata().getQueriesList());
+
+      // Update with document built from `documentSnapshot` because it is newer.
+      if (originalDocument == null
+          || documentSnapshot
+                  .getReadTime()
+                  .compareTo(Timestamp.fromProto(originalDocument.getMetadata().getReadTime()))
```

Comment:
This call to compare to should guard against `getReadTime()` returning null, since there doesn't appear to be a previous check ensuring it's non-null.

Reply:
I gave it a try, and it looks awkward. I am inclining to not check it here, because the only place setting `originalDocument` is below, and the need to defend against that is not very strong IMHO.

Reply:
The `compareTo` method on Timestamp expects a non-null argument (and doesn't assert it unfortunately) and there is the possibility that `getReadTime()` returns a null value, so we do in fact need to guard this check so that it doesn't result in an exception.

A similar scenario is present below on line 88 (in this commit) where it's possible a null value is returned before calling `toProto()`.

Reply:
Done.